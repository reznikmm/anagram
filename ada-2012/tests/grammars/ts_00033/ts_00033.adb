--  Check local attributes in Ordered AG
with Gela.Grammars.Reader;
with Gela.Grammars.Ordered;
with Gela.Grammars_Checks;
with Ada.Wide_Wide_Text_IO;

procedure TS_00033 is
   G : constant Gela.Grammars.Grammar :=
     Gela.Grammars.Reader.Read ("test.ag");
   Order  : Gela.Grammars.Ordered.Order_Maps.Map;
   Partitions : Gela.Grammars.Ordered.Partition_Array (G.Declaration'Range);

   use Gela.Grammars;

   Found : Boolean;
begin
   if not Gela.Grammars_Checks.Is_Well_Formed (G, True) then
      raise Constraint_Error;
   end if;

   Ordered.Find_Order (G, Found, Partitions, Order);

   declare
      use Gela.Grammars.Ordered.Order_Maps;
      Pos : Cursor := Order.First;
   begin
      while Has_Element (Pos) loop
         declare
            K : constant Ordered.Key := Key (Pos);
            V : constant Ordered.Action := Element (Pos);
            Prod : Production renames G.Production (K.Prod);
         begin
            Ada.Wide_Wide_Text_IO.Put
              (G.Non_Terminal (Prod.Parent).Name.To_Wide_Wide_String);
            Ada.Wide_Wide_Text_IO.Put (".");
            Ada.Wide_Wide_Text_IO.Put (Prod.Name.To_Wide_Wide_String);
            Ada.Wide_Wide_Text_IO.Put (Natural'Wide_Wide_Image (K.Pass));
            case V.Kind is
               when Ordered.Evaluate_Rule =>
                  Ada.Wide_Wide_Text_IO.Put (" eval: ");
                  Ada.Wide_Wide_Text_IO.Put
                    (G.Rule (V.Rule).Text.To_Wide_Wide_String);
               when Ordered.Descent =>
                  Ada.Wide_Wide_Text_IO.Put (" descent: ");
                  Ada.Wide_Wide_Text_IO.Put
                    (G.Part (V.Part).Name.To_Wide_Wide_String);
                  Ada.Wide_Wide_Text_IO.Put
                    (Natural'Wide_Wide_Image (V.Pass));
            end case;

            Ada.Wide_Wide_Text_IO.New_Line;
            Next (Pos);
         end;
      end loop;
   end;

end TS_00033;
