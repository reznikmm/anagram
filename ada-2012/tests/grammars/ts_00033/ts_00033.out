A.B 1 eval:  ${B.env} := 1; 
A.B 1 descent: B 1
A.B 1 eval:  ${A.value} := ${B.value}; 
B.C 1 eval:  ${C.env} := ${B.env}; 
B.C 1 descent: C 1
B.C 1 eval:  ${tmp} := ${C.value}; 
B.C 1 eval:  ${B.value} := ${C.tmp}; 
C.name 1 descent: name 1
C.name 1 eval:  ${C.value} := ${name.symbol} + ${C.env}; 
name.id_token 1 eval:  ${name.symbol} := ${id_token.symbol}; 
